File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.py
def solution(sequence) -> int:
    length: int = len(sequence)
    pulse1: list[int] = [
        sequence[i] if i % 2 == 0 else -sequence[i] for i in range(length)
    ]
    pulse2: list[int] = [
        sequence[i] if i % 2 != 0 else -sequence[i] for i in range(length)
    ]
    
    def max_sub_sum(nums) -> int:
        max_sum: int = nums[0]
        current_sum: int = nums[0]
        
        for i in range(1, len(nums)):
            num: int = nums[i]
            current_sum = max(num, current_sum + num)
            max_sum = max(max_sum, current_sum)
        return max_sum
    
    answer: int = max(max_sub_sum(pulse1), max_sub_sum(pulse2))
    
    return answer

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
class Solution {
    public long solution(int[] sequence) {
        int n = sequence.length;
        long answer = Long.MIN_VALUE;
        long dp1 = 0;
        long dp2 = 0;
        for (int i = 0; i < n; i++) {
            int val1 = (i % 2 == 0 ? 1 : -1) * sequence[i];
            int val2 = (i % 2 == 0 ? -1 : 1) * sequence[i];
            dp1 = Math.max(val1, dp1 + val1);
            dp2 = Math.max(val2, dp2 + val2);
            answer = Math.max(answer, Math.max(dp1, dp2));
        }
        return answer;
    }
}